### PR TITLE
Vms straight-line-pass implementation

### PR DESCRIPTION
**Pull request by marques-valeria**

### CODE REVIEW

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    private boolean hasSameLineNumber(String classInfo, int oldLine, int newLine) {
+        for (String className : lineChanges.keySet()) {
+            if (className.contains(classInfo)) {
+                int offset = 0;
+                for (Integer originalLine : lineChanges.get(className).keySet()) {
+                    if (originalLine <= oldLine) {
+                        offset += lineChanges.get(className).get(originalLine);
+                    }
+                }
+                if (newLine - oldLine <= offset) {
```
Comment (bcarlet) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:220]: It's likely that I'm misunderstanding how this works, but I'm skeptical of this line. If all modifications were in place, then `offset` should be 0. In which case, we'd be returning `true` whenever `newLine <= oldLine`, which  doesn't seem right.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+    }
+
+    /**
+     * Rewrites violation-counts to only include violations in newViolations.
+     */
+    private void rewriteViolationCounts() throws MojoExecutionException {
+        // for each line of violation-counts, if it can be mapped to a new violation it gets to stay (else it goes)
+        try {
+            List<String> lines = Files.readAllLines(new File(".starts/violation-counts").toPath());
+            PrintWriter writer = new PrintWriter(".starts/violation-counts");
```
Comment (bcarlet) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:237]: Same comment as above with regard to `.starts`.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+        saveViolationCounts();
+        findLineChangesAndRenames(getDiffs());
+        getLog().info("Number of files renamed: " + renames.size());
+        getLog().info("Number of changed files found: " + lineChanges.size());
+        oldViolations = parseViolations(".starts/violation-counts-old");
+        newViolations = parseViolations(".starts/violation-counts");
+        getLog().info("Number of total violations found: " + newViolations.size());
+        removeDuplicateViolations();
+        getLog().info("Number of unique violations found: " + newViolations.size());
+        rewriteViolationCounts();
```
Comment (bcarlet) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:65]: It seems like this would have a cyclic effect where violations that aren't fixed show up on every other run. Is that the intended behavior?

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 1 above)
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java]: Yeah this only works for insertions, I updated the logic to account for both insertions and deletions

File: emop-maven-plugin/pom.xml [hunk 4]
```diff
@@ -44,20 +49,6 @@
         <configuration>
           <skipErrorNoDescriptorsFound>true</skipErrorNoDescriptorsFound>
         </configuration>
-        <executions>
```
[the remark this answers is not in the document]
Comment (marques-valeria) [reply emop-maven-plugin/pom.xml:47]: We removed this while trying to debug JGit's installation - my understanding is that it mostly checks for style but I can do more experimentation to see if this is really necessary or not

[the remark this answers is not in the document]
Comment (marques-valeria) [reply emop-maven-plugin/pom.xml]: Yeah this change doesn't seem essential, I reversed it for consistency with previous versions of emop

File: emop-core/src/main/java/edu/cornell/emop/util/Util.java [hunk 5]
```diff
@@ -102,4 +103,66 @@ public static Set<String> retrieveProjectPackageNames(File classesDir) {
...
+                    .stream()
+                    .map(Util::parseViolation)
+                    .collect(Collectors.toSet());
+        } catch (IOException exception) {
+            return new HashSet<>();
+        }
+    }
+
+    /**
+     * Analyzes a violations file and returns a list of violations.
```
Comment (ayakayorihiro) [inline emop-core/src/main/java/edu/cornell/emop/util/Util.java:125]: I think you might want to reword this to say: Analyzes a violations file and returns the set of specifications that were violated.

File: emop-maven-plugin/src/main/java/edu/cornell/RVMCleanMojo.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.io.File;
+import java.nio.file.Paths;
+
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.surefire.SurefirePlugin;
+import org.apache.maven.plugins.annotations.Mojo;
+
+@Mojo(name = "clean-rvm", requiresDirectInvocation = true)
+public class RVMCleanMojo extends SurefirePlugin {
```
Comment (ayakayorihiro) [inline emop-maven-plugin/src/main/java/edu/cornell/RVMCleanMojo.java:11]: why do we need to extend the SurefirePlugin here?

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    //          file        line     lines modified
+    private Map<String, Map<Integer, Integer>> lineChanges = new HashMap<>();
+
+    // Represents the renamed file's original name
+    private Map<String, String> renames = new HashMap<>();
+
+    // Describes new classes which have been made between commits
+    private Set<String> newClasses = new HashSet<>();
+
+    private Set<List<String>> oldViolations;
```
Comment (ayakayorihiro) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:55]: Not super necessary but a good to have would be to briefly comment on the indices of each List (it seems like a triple between spec, class, line number?)

Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:55]: or you could even refactor to create a new class `Violation`

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    /**
+     * Rewrites violation-counts to only include violations in newViolations.
+     */
+    private void rewriteViolationCounts() throws MojoExecutionException {
+        // for each line of violation-counts, if it can be mapped to a new violation it gets to stay (else it goes)
+        try {
+            Path vc = Paths.get(System.getProperty("user.dir"), "violation-counts");
+            List<String> lines = Files.readAllLines(vc);
+            PrintWriter writer = new PrintWriter(vc.toFile());
+            for (String line : lines) {
```
Comment (ayakayorihiro) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:227]: I feel like you can optimize the whole routine by making a `Violation` class, and having a `printViolation()` method there that outputs the usual error message that you'd see in `violation-counts`. That way, you don't have to reread the File, you can just overwrite whatever was there with messages from the set of new specs you have.

**Review by ayakayorihiro — CHANGES_REQUESTED** [review]
Great work, this mostly looks great! I think your `VmsMojo` would benefit greatly from having/interfacing with a class called `Violation` (or whatever you'd like to call it) that has spec, class, line number info as field variables instead of the `List<String>` setup you have right now.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 8 above)
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:227]: I think you're right that we can reconstruct the error message from a violation (we would have to remember the count at the beginning as well as the function the violation occurred, along with everything else) if I have extra time this week I can look into optimizing it but I don't think it's high on the priority list - do you think you could maybe raise an issue?

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 7 above)
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:55]: Made a new `Violation` class :)

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 8 above)
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:227]: Sure thing, I can make an issue on this! This is definitely more of an optimization thing and we should prioritize getting all of the features out there :)

**Review by ayakayorihiro — APPROVED** [review]

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+                    }
+                }
+            }
+        } catch (IOException exception) {
+            throw new MojoExecutionException("Encountered an error when comparing different files");
+        }
+    }
+
+    /**
+     * Scrubs newViolations of violations believed to be duplicates from violation-counts-old.
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:153]: Scrubs --> Removes

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+                            newViolation.getLineNum()))
+                    .collect(Collectors.toSet());
+
+            if (!relevantOldViolations.isEmpty()) {
+                violationsToRemove.add(newViolation);
+            }
+        }
+        for (Violation violationToRemove : violationsToRemove) {
+            newViolations.remove(violationToRemove);
+        }
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:172]: Replace lines 170 to 172 with something `newViolations.removeAll(violationsToRemove);`

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    }
+
+    /**
+     * Scrubs newViolations of violations believed to be duplicates from violation-counts-old.
+     */
+    private void removeDuplicateViolations() {
+        Set<Violation> violationsToRemove = new HashSet<>();
+        for (Violation newViolation : newViolations) {
+            Set<Violation> relevantOldViolations = oldViolations.stream()
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:158]: is this relevantOldViolation always empty or a singleton set? That is, can it ever have more than one match?

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+                    .filter(oldViolation -> oldViolation.getSpecification().equals(newViolation.getSpecification()))
+                    .filter(oldViolation -> oldViolation.getClassInfo().equals(newViolation.getClassInfo())
+                                         || isRenamed(oldViolation.getClassInfo(), newViolation.getClassInfo()))
+                    .filter(oldViolation -> hasSameLineNumber(oldViolation.getClassInfo(), oldViolation.getLineNum(),
+                            newViolation.getLineNum()))
+                    .collect(Collectors.toSet());
+
+            if (!relevantOldViolations.isEmpty()) {
+                violationsToRemove.add(newViolation);
+            }
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:168]: I think you should be able to just `addAll` the result of the filtering to `violationsToRemove` without first checking for emptiness.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+        for (String renamedClass : renames.keySet()) {
+            if (renamedClass.contains(newClass) && renames.get(renamedClass).contains(oldClass)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether an old line in a file can be mapped to the new line. There's room for optimization.
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:192]: > There's room for optimization.

Make this into a TODO comment and say how you are thinking to do the optimization

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+                        offset += lineChanges.get(className).get(originalLine);
+                    }
+                }
+                if (newLine >= oldLine) { // if lines have been inserted
+                    return offset >= 0 && offset >= newLine - oldLine;
+                } else { // if lines have been removed
+                    return offset <= 0 && offset <= newLine - oldLine;
+                }
+            }
+        }
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:214]: I don't understand any of this code because I do not see an example. Can you add one to the comments?

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+        try {
+            Path vc = Paths.get(System.getProperty("user.dir"), "violation-counts");
+            List<String> lines = Files.readAllLines(vc);
+            PrintWriter writer = new PrintWriter(vc.toFile());
+            for (String line : lines) {
+                if (isNewViolation(line)) {
+                    writer.println(line);
+                }
+            }
+            writer.close();
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:232]: if you use try-with-resources, you will not need to close() by yourself

File: vms_design_document.md [hunk 16]
```diff
@@ -0,0 +1,8 @@
+# Violation Message Suppression (VMS) Implementation
```
Comment (owolabileg) [inline vms_design_document.md:1]: Why have a second design document?

**Review by owolabileg — CHANGES_REQUESTED** [review]
This is starting to take a good shape. 

Address the comments and let me know when I should take another look.

We should also plan to discuss some of the hard-to-understand code in person.

File: emop-core/src/main/java/edu/cornell/emop/util/Violation.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class Violation {
+
+    private String specification;
+    private String classInfo;
+    private int lineNum;
+
+    public Violation(String specification, String classInfo, int lineNum) {
```
[the remark this answers is not in the document]
Comment (marques-valeria) [reply emop-core/src/main/java/edu/cornell/emop/util/Violation.java:16]: Class is a keyword, so it can't just be named class

[the remark this answers is not in the document]
Comment (owolabileg) [reply emop-core/src/main/java/edu/cornell/emop/util/Violation.java:16]: True :) How about `className` (just so that it's clearer what is being referred to)

File: emop-maven-plugin/pom.xml [hunk 18]
```diff
@@ -28,6 +28,11 @@
       <artifactId>aspectjtools</artifactId>
       <version>1.8.1</version>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.jgit</groupId>
+      <artifactId>org.eclipse.jgit</artifactId>
+      <version>5.13.1.202206130422-r</version>
```
[the remark this answers is not in the document]
Comment (marques-valeria) [reply emop-maven-plugin/pom.xml:34]: I'm not sure what the difference between a major/minor version is, also is this not stable? I got it from the maven central repository

File: emop-core/src/main/java/edu/cornell/emop/util/Violation.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (bcarlet) [reply emop-core/src/main/java/edu/cornell/emop/util/Violation.java:16]: I think `class` is an invalid identifier name?

File: emop-maven-plugin/src/main/java/edu/cornell/RVMCleanMojo.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+import java.nio.file.Paths;
+
+import org.apache.maven.plugin.AbstractMojo;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.Mojo;
+
+@Mojo(name = "clean-rvm", requiresDirectInvocation = true)
+public class RVMCleanMojo extends AbstractMojo {
+    public void execute() throws MojoExecutionException {
+        File vc = Paths.get(System.getProperty("user.dir"), "violation-counts").toFile();
```
[the remark this answers is not in the document]
Comment (bcarlet) [reply emop-maven-plugin/src/main/java/edu/cornell/RVMCleanMojo.java:13]: Since RV-Monitor appends to `violation-counts` rather than overwriting, we need to clean it before each run of VMS.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+     * Updates the lineChanges and renames based on found differences.
+     *
+     * @param diffs List of differences between two versions of the same program
+     * @throws MojoExecutionException if error is encountered during runtime
+     */
+    private void findLineChangesAndRenames(List<DiffEntry> diffs) throws MojoExecutionException {
+        try {
+            for (DiffEntry diff : diffs) {
+                // If the old path is /dev/null, then the file has been created between commits
+                if (diff.getOldPath().equals("/dev/null")) {
```
[the remark this answers is not in the document]
Comment (bcarlet) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:128]: Yes, the method is defined to return `"/dev/null"` in certain cases regardless of OS. That said, jGit also defines a constant `DEV_NULL` that we can use to make things clearer.

[the remark this answers is not in the document]
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:128]: Under the hood, yes - I changed this line to DiffEntry.DEV_NULL to make it a little more agnostic

[the remark this answers is not in the document]
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:128]: Doc here: https://archive.eclipse.org/jgit/docs/jgit-2.0.0.201206130900-r/apidocs/org/eclipse/jgit/diff/DiffEntry.html

[the remark this answers is not in the document]
Comment (owolabileg) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:128]: Please add that doc to a comment in the code.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 11 above)
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:158]: Yes, if the same violation occurs at least twice in the old file within the offset then they will both be mapped to the same violations in the new file (within the offset)

Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:158]: Something I could do is change this filtered set to a for each loop -> if any of the old violations can be successfully mapped to the new one then we can add the new violation to the list of violations to remove, this would save us the cost of iterating through all old violations every time

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 12 above)
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:168]: This portion of the code has been reworked

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java (hunk 15 above)
Comment (marques-valeria) [reply emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:232]: Out of curiosity, why is this?

File: vms_design_document.md (hunk 16 above)
Comment (marques-valeria) [reply vms_design_document.md:1]: This was a holdover from when we were first thinking about vms, it's been removed now

File: emop-maven-plugin/src/main/java/edu/cornell/RVMCleanMojo.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (bcarlet) [reply emop-maven-plugin/src/main/java/edu/cornell/RVMCleanMojo.java]: We've updated the Mojo to also implement #25, with a mode to only clean `violation-counts`. That way we'll only have one clean Mojo.

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+     * Old code          New code
+     * line 1            line 1
+     * line 2            line 2
+     * line 3            new line
+     * line 4            line 3
+     *                   line 4
+     * Lines 1 and 2 do not see any differences, so their offset is 0 and they will only map to lines 1 and 2 in the new
+     * code, respectively. Line 3 sees an addition of one line, so the offset here is 1 and both the new line and line 3
+     * in the new code will map to the previous line 3. Similarly, line 4 also takes the previous addition of a line
+     * into account and have an offset of 1, and both lines 3 and 4 in the new code will map to the previous line 4.
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:217]: have an offset --> has an offset

File: emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java [hunk 22]
```diff
@@ -0,0 +1,313 @@
...
+     * Deletions work similarly, and are represented by a negative offset.
+     * TODO: The current implementation is generous with finding the "same" violation. The offset is very accurate in
+     *  finding the precise location unmodified code has been moved to. This information can be leveraged for more
+     *  accuracy when differentiating which violations are new or old when they occur close together and the line
+     *  itself where the violation occurred is not a part of any direct differences in code. There are also errors that
+     *  can arise because of differences between the last commit and the previous run of violation-counts (whose
+     *  violations we keep track of) this can be fixed by incorporating sha information.
```
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java:224]: Please also open an issue for this TODO

**Review by owolabileg — COMMENTED** [review]
Please address the super comments that I made, and then merge

File: emop-maven-plugin/src/main/java/edu/cornell/CleanMojo.java [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+     * Parameter to determine which artifacts to clean.
+     */
+    @Parameter(property = "clean.mode", defaultValue = "EVERYTHING")
+    private CleanMode mode;
+
+    public void execute() throws MojoExecutionException {
+        switch (mode) {
+            case EVERYTHING:
+                super.execute();
+                FileUtil.delete(Paths.get(System.getProperty("user.dir"), "weaved-bytecode").toFile());
```
[the remark this answers is not in the document]
Comment (bcarlet) [reply emop-maven-plugin/src/main/java/edu/cornell/CleanMojo.java:31]: `.starts` is removed by the STARTS clean Mojo in `super.execute()`.
